File: Sem2/ASD/SortingAlgorithms/quickSort.py
# Quick sort algorithm
#
# DESCRIPTION
# Quicksort is a divide-and-conquer algorithm. It works by selecting a 'pivot' element from the array
# and partitioning the other elements into two sub-arrays, according to whether they are less than or
# greater than the pivot. The sub-arrays are then sorted recursively.
# The 'pivot' can be chosen in many ways. I could be e.g. the first, the middle or the last element.
# By choosing the last element as the pivot we spare one swap operation each instance.
# 
# COMPLEXITY
# Average time complexity:  O(n log n)
# Pessimistic time complexity: O(n^2)
# Space complexity: *depends on the implementation*
# This one has space complexity O(1)

def quickSort(L, start, end):
    def partition (L, start, end):
        x = L[end]
        i = start-1
        for j in range(start, end):
            if L[j] <= x:
                i += 1
                L[i], L[j] = L[j], L[i]
        L[i+1], L[end] = L[end], L[i+1]
        return i+1

    if start < end:
        middle = partition(L,start,end)
        quickSort(L, start, middle-1)
        quickSort(L, middle+1, end)


L = [3,5,2,10,8,4,7,6,1,9]

print(L)

quickSort(L, 0, len(L)-1)

print(L)

File: Sem2/ASD/SortingAlgorithms/quickSort.cpp
#include <iostream>

using namespace std;

void swap(int* a, int* b)  
{  
    int t = *a;  
    *a = *b;  
    *b = t;  
}

int partition(int tab[], int p, int r) {
    int x = tab[r], i = p - 1;
    for( int j = p; j < r; j += 1) {
        if( tab[j] <= x ) {
            i += 1;
            swap(&tab[i], &tab[j]);
        }
    }
    swap(&tab[i+1], &tab[r]);
    return i+1;
}

void quickSort(int tab[], int p, int r) {
    if( p < r ) {
        int q = partition(tab, p, r);
        quickSort(tab, p, q-1);
        quickSort(tab, q+1, r);
    }
}

void printArray(int tab[]) {
    for (int i = 0; i < 15; i += 1) {
        cout << tab[i] << " ";
    }
    cout << '\n';
}

int main() {
    int tab[15] = {1, 2, 99, 22, 542, 54, 10, 3, 4, 4, 4, 2, 5, 1, 1};
    printArray(tab);
    quickSort(tab, 0, 14);
    printArray(tab);
    return 0;
}